Fix bugs, compiler warnings and add tests

birktj opened the pull request:
This PR adds a `Limits` struct to fix the memory exhaustion bug in #29, this changes the public API slightly, because `ifd::Entry::val` now also needs a `Limits` parameter. It adds some array bounds checking for indices that directly come from the tiff file. While I was at it I also fixed #13 by changing `Decoder::read_string` to return a `TiffResult` instead of a `FromUtf8Error` result. I also fixed some smaller compiler and clippy warnings. 

I added the samples from #28, #29 and #31 as new tests, thank you @Shnatsel!

This will fix #13, fix #28, fix #29, fix #30 and fix #31.

@Robzz and @HeroicKatora what do you think of this?

Shnatsel commented on src/decoder/mod.rs:
```diff
@@ -95,13 +126,36 @@ struct StripDecodeState {
+    decoding_buffer_size: u64,
+    /// The maximum size of any ifd value in bytes, the default is 
+    /// 1MiB.
+    ifd_value_size: u64,
```
Any particular reason why this is `u64` and not `usize`? You cannot really have a u64-sized buffer on a 32-bit platform.

birktj replied:
Not really, I just took from the `Limits` struct in the `png` crate, but you are right `usize` would be a better fit.

Shnatsel commented:
Thanks for fixing these so quickly! This has unblocked further fuzzing which found at least one more panic: #33

HeroicKatora commented on src/decoder/mod.rs:
```diff
@@ -28,7 +27,25 @@ pub enum DecodingResult {
+    }
+
+    fn new_u16(size: usize, limits: &Limits) -> TiffResult<DecodingResult> {
+        if size as u64*2 > limits.decoding_buffer_size {
```
This one should probably be `size as u64 > limits.decoding_buffer_size/2` otherwise the mult may overflow, or use `checked_mul` instead.

HeroicKatora replied:
Overall, I'm not sure if the case of `usize == u32` is handled properly for all the multiplications everywhere (i.e. doesn't overflow them).

birktj commented:
@HeroicKatora I have fixed the possibly overflowing multiplication and changed the limits to `usize`. Do you think I need to change anything else or are we good to go?

birktj commented:
@HeroicKatora @Robzz @bvssvni friendly ping. Would it be possible to look at this so that we can eventually release a new version of the library with the newly merged PRs?

HeroicKatora commented:
Just now realized that during converting to `TiffResult`, the line you changed here https://github.com/PistonDevelopers/image-tiff/pull/32/files#diff-ee033e65f37f7717cb17dc37bc71de6bL301 was actually a bug as the error was ignored. Good job :+1: We should probably do a stability release sometime soon. Want to merge this today if no further concerns appear.
